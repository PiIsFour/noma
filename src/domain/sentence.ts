import { Word } from './word'

export enum CarriageType {
	ga = 'ga',
}

export interface Carriage<T = Word> {
	word: T,
	type: CarriageType,
	modifier?: Sentence<T>,
}

export const ga = <T = Word>(word: T, modifier?: Sentence<T>): Carriage<T> => ({ word, type: CarriageType.ga, modifier })

export enum EngineType {
	verb = 'verb',
	iAdj = 'iAdj',
}

export interface Engine<T = Word> {
	word: T,
	type: EngineType,
}

export const verb = <T = Word>(word: T): Engine<T> => ({ word, type: EngineType.verb })
export const iAdj = <T = Word>(word: T): Engine<T> => ({ word, type: EngineType.iAdj })

export interface Sentence<T = Word> {
	parts: Carriage<T>[],
	end: Engine<T>,
}
