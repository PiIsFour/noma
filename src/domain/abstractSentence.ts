import { Sentence } from './sentence'
import { PartsOfSpeech } from './word'

interface AbstractWord {
	conditions: ((p: PartsOfSpeech[]) => boolean)[]
}

export type AbstractSentence = Sentence<AbstractWord>
