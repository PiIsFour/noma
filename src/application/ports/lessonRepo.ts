import { AbstractSentence } from '../../domain/abstractSentence'

interface LessonBase {
	id: number
	level: number
}

export interface StandAloneLesson extends LessonBase {
	type: 'StandAloneLesson'
	sentence: AbstractSentence
}

export type Lesson = (StandAloneLesson)

export interface LessonRepo {
	getLessons(): Array<Lesson>
}
