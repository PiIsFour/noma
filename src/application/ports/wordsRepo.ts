import { Option } from 'fp-ts/Option'
import { PartsOfSpeech, Word } from '../../domain/word'

export interface WordRepo {
	getWord(filter: (partsOfSpeech: PartsOfSpeech[]) => boolean): Option<Word>
}
