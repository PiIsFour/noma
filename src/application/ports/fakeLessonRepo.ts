import { AbstractWord } from '../../domain/abstractSentence'
import { ga, iAdj, verb } from '../../domain/sentence'
import { PartsOfSpeech } from '../../domain/word'
import { Lesson, LessonRepo } from './lessonRepo'

const lessons: Lesson[] = []

export const fakeLessonRepo: LessonRepo = {
	getLessons: () => lessons,
}

export const createFakeLessonRepo = (lessons: Lesson[]): LessonRepo => ({
	getLessons: () => lessons,
})

const addLesson = (lesson: Omit<Lesson, 'id' | 'level'>): Lesson => {
	const newLesson = {
		...lesson,
		level: 0,
		id: lessons.length + 1,
	}
	lessons.push(newLesson)
	return newLesson
}

const aNoun: AbstractWord = {
	conditions: [PoS => PoS.includes(PartsOfSpeech.noun)],
}

const aVerb: AbstractWord = {
	conditions: [PoS => PoS.includes(PartsOfSpeech.verb)],
}

const aIAdj: AbstractWord = {
	conditions: [PoS => PoS.includes(PartsOfSpeech.iAdj)],
}

export const NounGaVerbLesson = addLesson({
	type: 'StandAloneLesson',
	sentence: {
		parts: [ga(aNoun)],
		end: verb(aVerb),
	},
})

export const NounGaIAdjLesson = addLesson({
	type: 'StandAloneLesson',
	sentence: {
		parts: [ga(aNoun)],
		end: iAdj(aIAdj),
	},
})
