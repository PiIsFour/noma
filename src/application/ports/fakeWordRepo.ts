import { fromNullable } from 'fp-ts/Option'
import { PartsOfSpeech, Word } from '../../domain/word'
import { WordRepo } from './wordsRepo'

export const sakura: Word = {
	hiragana: 'さくら',
	kanji: 'さくら',
	translation: 'sakura',
	partsOfSpeech: [PartsOfSpeech.noun],
}

export const walk: Word = {
	hiragana: 'あるく',
	kanji: '歩く',
	translation: 'walk',
	partsOfSpeech: [PartsOfSpeech.verb],
}

export const cute: Word = {
	hiragana: 'かわいい',
	kanji: '可愛い',
	translation: 'cute',
	partsOfSpeech: [PartsOfSpeech.iAdj],
}

export const createFakeWordRepo = (words: Word[]): WordRepo => ({
	getWord: filter => fromNullable(words.find(w => filter(w.partsOfSpeech))),
})
