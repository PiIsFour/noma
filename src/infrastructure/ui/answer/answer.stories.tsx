import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import { Answer } from './answer'
import { ga, verb } from '../../../domain/sentence'
import { sakura, walk } from '../../../application/ports/fakeWordRepo'

export default {
	title: 'Scenes/Answer',
	component: Answer,
	argTypes: {
	},
} as ComponentMeta<typeof Answer>

const Template: ComponentStory<typeof Answer> = (args) => <Answer {...args} />

export const Standard = Template.bind({})
Standard.args = {
	kanji: 'さくらが歩く。',
	sentence: {
		parts: [ga(sakura)],
		end: verb(walk),
	},
}
