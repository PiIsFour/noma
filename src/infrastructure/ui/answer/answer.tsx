import React from 'react'
import { Question } from '../../../application/useCase/getQuestion'
import { Carriage } from '../carriage/carriage'
import { Engine } from '../engine/engine'
import './answer.css'

export const Answer = ({ sentence }: Question): JSX.Element => {
	return <div className="answer">
		{sentence.parts.map((carriage, index) => <Carriage key={index} carriage={carriage} />)}
		<Engine engine={sentence.end} />
	</div>
}
