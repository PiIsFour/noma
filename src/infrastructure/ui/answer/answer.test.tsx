import React from 'react'
import { shallow } from 'enzyme'
import { ga, verb } from '../../../domain/sentence'
import { sakura, walk } from '../../../application/ports/fakeWordRepo'

import { Answer } from './answer'

describe('Answer', () => {
	it('matches snapshot', () => {
		const sentence = {
			parts: [ga(sakura)],
			end: verb(walk),
		}
		const wrapper = shallow(<Answer kanji='さくらが歩く。' sentence={sentence} />)
		expect(wrapper).toMatchSnapshot()
	})
})
