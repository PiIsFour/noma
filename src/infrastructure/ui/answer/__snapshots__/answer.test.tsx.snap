// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer matches snapshot 1`] = `
<div
  className="answer"
>
  <Carriage
    carriage={
      Object {
        "modifier": undefined,
        "type": "ga",
        "word": Object {
          "hiragana": "さくら",
          "kanji": "さくら",
          "partsOfSpeech": Array [
            "noun",
          ],
          "translation": "sakura",
        },
      }
    }
  />
  <Engine
    engine={
      Object {
        "type": "verb",
        "word": Object {
          "hiragana": "あるく",
          "kanji": "歩く",
          "partsOfSpeech": Array [
            "verb",
          ],
          "translation": "walk",
        },
      }
    }
  />
</div>
`;
